File: backend/src/main/kotlin/com/dotcom/retail/domain/order/Order.kt
package com.dotcom.retail.domain.order

import com.dotcom.retail.common.model.AuditingEntity
import com.dotcom.retail.common.model.Contact
import com.dotcom.retail.domain.user.User
import jakarta.persistence.*
import java.math.BigDecimal
import java.util.*

@Entity
@Table(name = "orders")
class Order(
    @Id
    val id: UUID = UUID.randomUUID(),

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    var user: User? = null,

    var sessionId: String? = null,
    var intentId: String,

    @Enumerated(EnumType.STRING)
    var status: OrderStatus = OrderStatus.PENDING_PAYMENT,

    @Embedded
    var contact: Contact? = null,

    var notes: String? = null,

    @Enumerated(EnumType.STRING)
    var shippingType: ShippingType? = ShippingType.STANDARD,

    var shippingCost: BigDecimal? = null,

    @Column(nullable = false)
    var totalAmount: BigDecimal,

    @OneToMany(mappedBy = "order", cascade = [CascadeType.ALL], orphanRemoval = true)
    var items: MutableList<OrderItem> = mutableListOf(),

    ) : AuditingEntity() {

    fun addItem(item: OrderItem) {
        items.add(item)
        item.order = this
    }

}

enum class OrderStatus {
    PENDING_PAYMENT,
    PAID,
    FAILED,
    SHIPPED,
    DELIVERED,
    CANCELLED
}

enum class ShippingType {
    STANDARD,
    EXPRESS
}

File: frontend/src/types/order.ts
import {AddressFields, Contact} from "@_types/contact"
import {type Page, PageResponse} from "@_types/page"

export interface Order {
    id: string,
    status: OrderStatus,
    paymentId: string,
    sessionId?: string,
    items: OrderItem[],
    shippingType: ShippingType,
    shippingCost: number,
    totalAmount: number,
    contact: Contact,
    date: number
}

export interface OrderItem {
    productId: string,
    productName: string,
    imageUrl: string,
    quantity: number,
    price: number,
    totalAmount: number,
}

export interface CreateOrderResponse {
    id: string
    status: string
    clientSecret: string
}

export interface SubmitOrderRequest {
    name: string
    email: string
    phone: string
    address: AddressFields,
    shippingType: ShippingType
    notes?: string
}

export interface OrderResponse {
    content: Order[]
    page: PageResponse
}

export enum OrderStatus { PENDING_PAYMENT, PAID, SHIPPED, DELIVERED, CANCELLED }
export enum ShippingType { STANDARD = "STANDARD", EXPRESS = "EXPRESS"}

export interface OrderParams extends Page {
    status: OrderStatus
    sort: "desc" | "asc"
}